---
title: "Heterogeneity analysis I"
output: html_notebook
---

To check out the code/data for this tutorial, run "git clone https://github.com/hms-dbmi/scw.git", and change into "mit2017/scde" folder.

## Load Camp et al. data
A Fluidigm C1 dataset from (Camp et al.)[http://www.pnas.org/content/112/51/15672] has been pre-prepared for you. The data is provided as a matrix of gene counts, where each column corresponds to a cell and each row a gene.

```{r}
data.dir <- "~/workshop_materials/transcriptomics/scw/scw2016/data/"
load(paste0(data.dir,"cd.RData"))
dim(cd)
cd[1:5,1:5]
```

Let's filter cells a bit:
```{r}
# filter out low-gene cells (often empty wells)
cd <- cd[, colSums(cd>0)>1.8e3]
# remove genes that don't have many reads
cd <- cd[rowSums(cd)>10, ]
# remove genes that are not seen in a sufficient number of cells
cd <- cd[rowSums(cd>0)>5, ]

# how many genes and cells after filtering?
dim(cd)
```

Look at the count distributions:
```{r fig.width=8,fig.height=4}
par(mfrow=c(1,2), mar = c(3.5,3.5,2.5,2.5), mgp = c(2,0.65,0), cex = 0.85)
hist(log10(colSums(cd)+1),main='reads per cell',col='wheat')
hist(log10(rowSums(cd)+1),main='reads per gene',col='wheat')
```


In the original publication, the authors proposed two main subpopulations: neurons and neuroprogenitor cells (NPCs). These labels have also been provided to you as a reference so we can see how different methods perform in recapitulating these labels.

```{r}
load(paste0(data.dir,'sg.RData'))
head(sg, 5)
```


## PCA
Note that there are over 10,000 genes that can be used to cluster cells into subpopulations. One fast, easy, and common technique to identify subpopulations is by using dimensionality reduction to summarize the data into 2 dimensions and then visually identify obvious clusters. Principal component analysis (PCA) is a linear dimensionality reduction method.

```{r fig.height=5,fig.width=5}
# define a log-transformed matrix (for poor man's variance stabilization)
mat <- log10(as.matrix(cd)+1)
# use principal component analysis for dimensionality reduction
base.pca <- prcomp(t(mat))
# visualize in 2D the first two principal components and color by cell type
par(mfrow=c(1,1), mar = c(3.5,3.5,2.5,2.5), mgp = c(2,0.65,0), cex = 0.85)
plot(base.pca$x[,1], base.pca$x[,2], col=sg, pch=16, main='PCA',xlab='PC1',ylab='PC2')
```

## t-SNE
T-distributed stochastic neighbor embedding (tSNE) is a non-linear dimensionality reduction method. Note that in tSNE, the perplexity parameter is an estimate of the number of effective neighbors. Here, we have 224 cells. A perplexity of 10 is suitable. For larger or smaller numbers of cells, you may want to increase the perplexity accordingly.

```{r}
library(Rtsne)
d <- dist(t(mat))
set.seed(0) # tsne has some stochastic steps (gradient descent) so need to set random 
tsne_out <- Rtsne(d, is_distance=TRUE, perplexity=10, verbose = TRUE, num_threads=1) 
```
Now let's plot it:
```{r fig.height=5,fig.width=5}
par(mar = c(3.5,3.5,2.5,2.5), mgp = c(2,0.65,0), cex = 0.85)
plot(tsne_out$Y, col=sg, pch=16, main='tSNE',xlab='',ylab='')
```


Note that with tSNE your results are stochastic:
```{r fig.height=5,fig.width=10}
set.seed(3) # tsne has some stochastic steps (gradient descent) so need to set random 
tsne_out2 <- Rtsne(d, is_distance=TRUE, perplexity=10, verbose = FALSE, num_threads=1) 
par(mfrow=c(1,2), mar = c(3.5,3.5,2.5,2.5), mgp = c(2,0.65,0), cex = 0.85)
plot(tsne_out$Y, col=sg, pch=16, main='tSNE #1',xlab='',ylab='')
plot(tsne_out2$Y, col=sg, pch=16, main='tSNE #2',xlab='',ylab='')
```

In general, the clusters from these tSNE results are not particularly clear-cut. Still, we may be wondering what genes and pathways characterize these subpopulation? For that, additional analysis is often needed and dimensionality reduction alone does not provide us with such insight.

## Pathway and gene set overdispersion analysis (PAGODA)
Alternatively, we may be interested in finer, potentially overlapping/non-binary subpopulations. For example, if we were clustering apples, PCA might separate red apples from green apples, but we may be interested in sweet vs. sour apples, or high fiber apples from low fiber apples. PAGODA is a method developed by the Kharchenko lab that enables identification and characterization of subpopulations in a manner that resolves these overlapping aspects of transcriptional heterogeneity. For more information, please refer to the original manuscript by (Fan et al.)[http://www.nature.com/nmeth/journal/v13/n3/full/nmeth.3734.html] PAGODA functions are implemented as part of the scde package.

```{r}
library(scde)
```

Each cell is modeled using a mixture of a negative binomial (NB) distribution (for the amplified/detected transcripts) and low-level Poisson distribution (for the unobserved or background-level signal of genes that failed to amplify or were not detected for other reasons). These models can then be used to identify robustly differentially expressed genes.

```{r eval=FALSE}
knn <- knn.error.models(cd, k = ncol(cd)/4, n.cores = 30, min.count.threshold = 2, min.nonfailed = 5, max.model.plots = 10)
```

```{r}
# just load from what we precomputed for you
load(paste0(data.dir,'knn.RData')) 
head(knn)
```

The cell specific error models are then used to quantify excess variance or overdispersion in genes and gene sets in order to cluster cells and identify subpopulations. Accurate quantification of this overdispersion means that we must normalize out expected levels of technical and intrinsic biological noise. Intuitively, lowly-expressed genes are often more prone to drop-out and thus may exhibit large variances simply due to such technical noise.

```{r}
varinfo <- pagoda.varnorm(knn, counts = cd, trim = 3/ncol(cd), max.adj.var = 5, n.cores = 2, plot = TRUE, verbose=TRUE)
```


It's usually a good idea to regress out the sequencing depth, as it commonly ends up distinguishing cells even after size normalization.
```{r}
# normalize out sequencing depth as well
varinfo <- pagoda.subtract.aspect(varinfo, colSums(cd[, rownames(knn)]>0))
```


Note: above calculations take some time so you can load a pre-calculated result:
```{r eval=FALSE}
# just load from what we precomputed for you
load(paste0(data.dir,'varinfo.RData'))
```


When assessing for overdispersion in gene sets, we can take advantage of pre-defined pathway gene sets such as GO annotations and look for pathways that exhibit statistically significant excess of coordinated variability. Intuitively, if a pathway is differentially perturbed, we expect all genes within said pathway to be upregulated or downregulated in the same group of cells. In PAGODA, for each gene set, we tested whether the amount of variance explained by the first principal component significantly exceed the background expectation.

```{r eval=FALSE}
library(org.Mm.eg.db)
# Translate gene names to ids
ids <- unlist(lapply(mget(rownames(cd), org.Mm.egALIAS2EG, ifnotfound = NA), function(x) x[1]))
# Reverse map
rids <- names(ids)
names(rids) <- ids
# Convert ids per GO category to gene names
go.env <- eapply(org.Mm.egGO2ALLEGS, function(x) as.character(na.omit(rids[x])))
go.env <- clean.gos(go.env) # Remove GOs with too few or too many genes
go.env <- list2env(go.env)  # Convert to an environment
```

Or use pre-compiled gene sets:
```{r}
# load gene sets
load(paste0(data.dir,'go.env.RData'))
# look at some gene sets
head(ls(go.env))
```

```{r}
# look at genes in gene set
get("GO:0000002 mitochondrial genome maintenance", go.env)
```

Do some filtering:
```{r}
# filter out gene sets that are too small or too big
go.env <- list2env(clean.gos(go.env, min.size=10, max.size=100))
# how many pathways
length(go.env)
```

Now, we can calculate weighted first principal component magnitudes for each GO gene set in the provided environment.
```{r eval=FALSE}
pwpca <- pagoda.pathway.wPCA(varinfo, go.env, n.components = 1, n.cores = 2)
```

In addition to relying on pre-defined pathways, we can also test on ‘de novo’ gene sets whose expression profiles are well-correlated within the given dataset.
```{r eval=FALSE}
clpca <- pagoda.gene.clusters(varinfo, trim = 7.1/ncol(varinfo$mat), n.clusters = 150, n.cores = 2, plot = FALSE)
```

But we'll use pre-calculated results, without de novo gene sets:

```{r}
load(paste0(data.dir,'pwpca.RData'))
clpca <- NULL # For the sake of time, set to NULL
```

We can now evaluate the statistical significance of the observed overdispersion for each GO gene set.
```{r}
df <- pagoda.top.aspects(pwpca, return.table = TRUE, plot = TRUE, z.score = 1.96)
```

Each point on the plot shows the PC1 variance (lambda1) magnitude (normalized by set size) as a function of set size. The red lines show expected (solid) and 95% upper bound (dashed) magnitudes based on the Tracey-Widom model.

```{r}
head(df)
```

* The z column gives the Z-score of pathway over-dispersion relative to the genome-wide model (Z-score of 1.96 corresponds to P-value of 5%, etc.).
* "z.adj" column shows the Z-score adjusted for multiple hypothesis (using Benjamini-Hochberg correction).
* "score" gives observed/expected variance ratio
* "sh.z" and "adj.sh.z" columns give the raw and adjusted Z-scores of "pathway cohesion", which compares the observed PC1 magnitude to the magnitudes obtained when the observations for each gene are randomized with respect to cells. When such Z-score is high (e.g. for GO:0008009) then multiple genes within the pathway contribute to the coordinated pattern.

These can be used to cluster cells:
```{r}
tam <- pagoda.top.aspects(pwpca, clpca, z.score = 1.96)
# determine overall cell clustering
hc <- pagoda.cluster.cells(tam, varinfo)
```


Because many of our annotated pathways and de novo gene sets likely share many genes or exhibit similar patterns of variability, we must reduce such redundancy to come up with a final coherent characterization of subpopulations.
```{r}
# reduce redundant aspects
tamr <- pagoda.reduce.loading.redundancy(tam, pwpca, clpca)
tamr2 <- pagoda.reduce.redundancy(tamr, plot = FALSE)
```

View the resulting aspects:
```{r}
pagoda.view.aspects(tamr2, cell.clustering = hc, box = TRUE, labCol = NA, margins = c(0.5, 20), col.cols = rbind(sg), top=10)
```

We can also use a 2D embedding on the resulting cell cell distances to aid visualization.
```{r}
# recalculate clustering distance .. we'll need to specify return.details=T
cell.clustering <- pagoda.cluster.cells(tam, varinfo, include.aspects=TRUE, verbose=TRUE, return.details=T)

# fix the seed to ensure reproducible results
set.seed(0)
tSNE.pagoda <- Rtsne(cell.clustering$distance, is_distance=TRUE, perplexity=10)

# plot
par(mfrow=c(1,1), mar = rep(5,4))
plot(tSNE.pagoda$Y, col=sg, pch=16, main='PAGODA tSNE')
```

By incorporating variance normalization and pathway-level information, our tSNE plot looks much more convincing!

We can also create an app to further interactively browse the results. A pre-compiled app has been launched for you (here)[http://pklab.med.harvard.edu/cgi-bin/R/rook/scw.xiaochang/index.html].

```{r}
app <- make.pagoda.app(tamr2,tam,varinfo,go.env,pwpca,col.cols=rbind(sg),cell.clustering = hc,title='neurons')
```

```{r}
show.app(app,'neurons',browse=F)
```


### Differential expression
Based on PAGODA results, we can see the main division between neurons and NPCs but we can also see further heterogeneity not visible by PCA or tSNE alone. In this case, prior knowledge on known marker genes can allow us to better interpret these identified subpopulations as IPCs, RGs, Immature Neurons, and Mature Neurons.

```{r fig.height=8}
# visualize a few known markers
markers <- c(
    "SCN2A","GRIK3","CDH6","NRCAM","SOX11",
    "SLC24A2", "SOX4", "DCX", "TUBB3","MAPT",
    "KHDRBS3",  "KHDRBS2", "KHDRBS1", "RBFOX3",
    "CELF6", "CELF5", "CELF4", "CELF3", "CELF2", "CELF1",
    "PTBP2", "PTBP1", "ZFP36L2",
    "HMGN2", "PAX6", "SFRP1",
    "SOX2", "HES1", "NOTCH2", "CLU","HOPX",
    "MKI67","TPX2",
    "EOMES", "NEUROD4","HES6"
    )
# heatmap for subset of gene markers
mat.sub <- varinfo$mat[markers,]
zlim <- c(-1,1)
mat.sub[mat.sub < zlim[1]] <- zlim[1]
mat.sub[mat.sub > zlim[2]] <- zlim[2]
heatmap(mat.sub[,hc$labels], Colv=as.dendrogram(hc), Rowv=NA, scale="none", col=colorRampPalette(c("blue", "white", "red"))(1024), ColSideColors=rainbow(2)[sg])
```

We can also define more refined subpopulations:
```{r fig.height=8}
sg2 <- as.factor(cutree(hc, k=4))
names(sg2) <- hc$labels
heatmap(mat.sub[,hc$labels], Colv=as.dendrogram(hc), Rowv=NA, scale="none", col=colorRampPalette(c("blue", "white", "red"))(1024), ColSideColors=rainbow(4)[sg2])
```


To further characterize identified subpopulations, we can identify differentially expressed genes between the two groups of single cells using scde. For more information, please refer to the original (manuscript)[http://www.nature.com/nmeth/journal/v11/n7/full/nmeth.2967.html].

First, let’s pick which identified subpopulations we want to compare using differential expression analysis.

```{r fig.height=8}
test <- as.character(sg2)
test[test==2] <- NA; test[test==3] <- NA
test <- as.factor(test)
names(test) <- names(sg2)
heatmap(mat.sub[,hc$labels], Colv=as.dendrogram(hc), Rowv=NA, scale="none", col=colorRampPalette(c("blue", "white", "red"))(1024), ColSideColors=rainbow(4)[test])
```

Now, let’s use scde to identify differentially expressed genes.
```{r}
# run differential expression tests on a subset of genes (to save time)
vi <- c("BCL11B", "CDH6", "CNTNAP2", "GRIK3", "NEUROD6", "RTN1", "RUNX1T1", "SERINC5", "SLC24A2", "STMN2", "AIF1L", "ANP32E", "ARID3C", "ASPM", "ATP1A2", "AURKB", "AXL", "BCAN", "BDH2", "C12orf48")
ediff <- scde.expression.difference(knn, cd[vi,], prior=varinfo$prior, groups = test, n.cores = 2, verbose = 1)
```

Look at the top genes:
```{r}
head(ediff[order(abs(ediff$Z), decreasing = TRUE), ], )
```

Look at a particular gene:
```{r fig.height=8, fig.width=5}
x <- scde.test.gene.expression.difference("NEUROD6", knn, cd, prior=varinfo$prior, groups = test)
```

Heatmap of differential genes:
```{r}
ediff.sig <- ediff[abs(ediff$cZ) > 1.96, ]
ediff.sig.up <- na.omit(rownames(ediff.sig[order(ediff.sig$cZ, decreasing = TRUE), ])[1:10])
ediff.sig.down <- na.omit(rownames(ediff.sig[order(ediff.sig$cZ, decreasing = FALSE), ])[1:10])
heatmap(mat[c(ediff.sig.up, ediff.sig.down), names(na.omit(test))], Rowv=NA, ColSideColors = rainbow(4)[test[names(na.omit(test))]],  col=colorRampPalette(c('blue', 'white', 'red'))(1024), scale="none")
```

Once we have a set of differentially expressed genes, we may use techniques such as gene set enrichment analysis (GSEA) to determine which pathways are differentially up or down regulated. GSEA is not specific to single cell methods and not included in this session but users are encouraged to check out this (light-weight R implementation with tutorials)[https://github.com/JEFworks/liger] on their own time.
